Closes #40

MatthewBaggins opened the pull request.

tayler6000 commented:
I'm not sure why this isn't linking correctly so, closes #40

Aprillion commented:
> I'm not sure why this isn't linking correctly so, closes #40

GH doesn't link from PR titles, only from PR comments (including description) and from commit messages 🤷

Aprillion commented on modules/stampcollection.py:
```diff
@@ -1,12 +1,28 @@
+        "stamp": StampType.RED,
+        "goldstamp": StampType.GOLD
+    }
+    )
```
IMHO these enums feel overcomplicated, there are 2 lookups: from emoji to enum, then from enum to vote_strength,

I would be more comfortable with just 1 lookup - fewer moving parts that can go wrong, only 1 place in the code to update when needed:
```py
vote_strengths_per_emoji = {
 "stamp": 1,
 "goldstamp": 5
}
```
and then:
```py
vote_strength = vote_strengths_per_emoji.get(emoji, 0)
```

MatthewBaggins replied:
I'm inclined to agree. I assumed that @UlisseMini meant to refactor strings in cases like these into `Enum`s. I'm a newbie to Stampy but I'm not sure in what cases should `vote_strength` be `0`? Intuitively, this doesn't make much sense, so maybe we could just change it into a boolean, `is_gold_stamp`?

tayler6000 replied:
Couldn't we just have the strength be a built in property of the Enum?

```py
@property
def vote_strength(self) -> int:
    return self.__value__
```

You would need to make sure to set the proper values in the declaration though.

You could also do

```py
def __int__(self) -> int:
    return self.vote_strength()
```

for an easy alternative option.

MatthewBaggins replied:
@tayler6000 Both of these throw errors

Aprillion replied:
there are thousands of emojis for Discord reactions (and whatever they call it on other supported clients/services):
![image](https://user-images.githubusercontent.com/1087670/190892998-196708fa-346a-4f8e-ba1c-c4176179d47d.png)

out of those, 2 have a special meaning: `stamp` provides a vote and `goldstamp`  provides 5x more more votes .. I still believe my suggested code is the best match for that.

MatthewBaggins replied:
I agree but IMO calling the argument `stamp_type` when it can be any emoji is confusing. I suggest either changing it to `emoji` or `reaction` and then doing it the way suggested or checking whether the emoji is a stamp before calling `update_vote`.

Aprillion replied:
oh, right, if it wasn't clear from my suggestion, I'm happy to rename the argument to `emoji` (or `reaction` if that will look better in your opinion)

MatthewBaggins replied:
I think I see now why it was done this way. `reset_stamps` calls `update_vote` with `stamp_type=god_id="0"` in order to recalculate stamps and do other updates. Maybe we should remove the part that does that (starting from `self.utils.users = self.utils.get_users()` AFAIK) to a separate method in order to keep one responsibility per method?

Aprillion commented on utilities/utilities.py:
```diff
@@ -348,19 +353,21 @@ def check_for_new_youtube_comments(self):
 
     def get_question(
-        self, order_type="TOP", wiki_question_bias=SemanticWiki.default_wiki_question_percent_bias
+        self, order_type: OrderType = OrderType.TOP, wiki_question_bias: float =SemanticWiki.default_wiki_question_percent_bias
```
anything wrong about type inference?

MatthewBaggins replied:
IMO it's a good practice to type hint arguments so that there's no ambiguity wrt their types. Maybe I'm excessively pedantic in that regard but from my experience this prevents a lot of runtime errors and increases code readability.

MatthewBaggins commented:
Sorry for recklessly merging from master and polluting "files changed".

@robertskmiles I deleted `process_reaction_event` methods
@Aprillion I replaced the enum with a emoji-to-vote-strength dict

Aprillion commented on config.py:
```diff
@@ -81,6 +81,7 @@ def __hash__(self):
 }[ENVIRONMENT_TYPE]
 
 bot_dev_channel_id = {"production": 808138366330994688, "development": 803448149946662923}[ENVIRONMENT_TYPE]
+error_channel_id = {"production": 1017527224540344380, "development": 1017531179664150608}[ENVIRONMENT_TYPE]
```
WTF is wrong with GitHub, this code is already in master branch, this is **not** how a PR diff is supposed to be shown after rebasing to master

please close this PR and open a new one, this is impossible to review in this shape IMHO

Aprillion replied:
@MatthewBaggins I'm sorry that GitHub was not able to support your workflow for updating this PR to master, if you know any way to review changes between your branch and master without closing this PR and opening a new one, let me know

MatthewBaggins replied:
Ok, opened a new one
